Connection to reverse interface dropped - UR5

### Affected ROS Driver version(s)

i do not know which，i download recently from here

### Used ROS distribution.

Noetic

### Which combination of platform is the ROS driver running on.

Linux

### How is the UR ROS Driver installed.

From binary packets

### Which robot platform is the driver connected to.

Real robot

### Robot SW / URSim version(s)

i do not know which

### How is the ROS driver used.

Through the robot teach pendant using External Control URCap

### Issue details

# Summary
My computer is Ubuntu 20.04, ROS-Noetic, dual system, and 5.15.0-113Generic kernel. After I enabled urcap in the teach pendant, I established a connection with the real robot through the driver. I use the control cabinet io output to control the external air pump and drive ur5 to execute the trajectory planned by moveit at the same time, but the problem occurs. Once the io signal and the planned trajectory are running at the same time, urcap will be interrupted and the sending signal will prompt failure:
```
[ INFO] [1720767234.769926758]: Completed trajectory execution with status PREEMPTED ...
[ INFO] [1720767234.772389043]: Execution completed: PREEMPTED
[ INFO] [1720767237.051394741]: Robot requested program
[ INFO] [1720767237.051487070]: Sent program to robot
[ INFO] [1720767237.401967035]: Robot connected to reverse interface. Ready to receive control commands.
[ INFO] [1720767257.018619063]: Connection to reverse interface dropped.
```
But I don't use io signals, and the execution of the track alone can be successful, but sometimes it will be interrupted. I don't know where the problem lies?

## Expected Behavior
At the same time, use the control cabinet io output to control the external air pressure pump and drive ur5 to execute the trajectory planned by moveit

## Actual Behavior
*The problem occurs. Once the io signal and the planned trajectory are running at the same time, urcap will be interrupted and the sending signal will prompt failure:
[ INFO] [1720767257.018619063]: Connection to reverse interface dropped.*



### Relevant log output

```shell
[ INFO] [1720767194.574030563]: Execution completed: FAILED
[ INFO] [1720767203.510599820]: Robot requested program
[ INFO] [1720767203.510698431]: Sent program to robot
[ INFO] [1720767203.881064833]: Robot connected to reverse interface. Ready to receive control commands.
[ INFO] [1720767204.136764422]: Connection to reverse interface dropped.
[ INFO] [1720767205.429355994]: Received request to compute Cartesian path
[ INFO] [1720767205.430741163]: Attempting to follow 31139 waypoints for link 'tool_ros' using a step of 0.010000 m and jump threshold 0.000000 (in global reference frame)
[ INFO] [1720767205.662247666]: Robot requested program
[ INFO] [1720767205.662335968]: Sent program to robot
[ INFO] [1720767206.017103343]: Robot connected to reverse interface. Ready to receive control commands.
[ INFO] [1720767206.092052969]: Computed Cartesian path with 31142 points (followed 100.000000% of requested trajectory)
[ INFO] [1720767206.335785831]: Execution request received
[ INFO] [1720767234.761974996]: Connection to reverse interface dropped.
```


### Accept Public visibility

- [X] I agree to make this context public

> I've put together a set of pull requests:
> 
> * [Add a service to set an analog output ros-industrial/ur_msgs#28](https://github.com/ros-industrial/ur_msgs/pull/28)
> * [Analog domain Universal_Robots_Client_Library#211](https://github.com/UniversalRobots/Universal_Robots_Client_Library/pull/211)
> * [Add a set_analog_output service #714](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/714)
> 
> @zp2546265641 If you could try that out and give some feedback, that would be great!
> 
> Please note: When building the client library from source, we suggest removing the installed apt package (`sudo apt remove ros-noetic-ur-client-library`)

hello, i write it as followed:
class URInterface:
    def __init__(self):
        self.set_io_client = rospy.ServiceProxy('/ur_hardware_interface/set_io', SetIO)
        self.set_analog_output_client = rospy.ServiceProxy('/ur_hardware_interface/set_io', SetAnalogOutput)

    def set_digital_io(self, num, state):
        request = SetIORequest()
        request.fun = SetIORequest.FUN_SET_DIGITAL_OUT
        request.pin = num
        request.state = state
        self.set_io_client(request)
        # self.set_io_client(SetIORequest.FUN_SET_DIGITAL_OUT, num, state)

    def set_analog_io(self, channel, voltage):
        request = SetAnalogOutputRequest()
        request.data.pin = channel
        request.data.domain = ur_msgs.msg.Analog.VOLTAGE  # 设置电压, 电流为CURRENT
        request.data.state = voltage
        # self.set_io_client(SetIORequest.FUN_SET_ANALOG_OUT, channel, voltage)
        self.set_analog_output_client(request)

    def close_pressure_system(self):
        rospy.loginfo("Closing pressure system...")
        self.set_analog_io(1,0)

    def open_pressure_system(self):
        rospy.loginfo("Opening pressure system...")
        self.set_digital_io(7, True)
        rospy.sleep(0.5)
        self.set_analog_io(1, 0.2)
if __name__ == '__main__':
    try:
        ur_interface = URInterface()
        # ur_interface.close_pressure_system()
        ur_interface.open_pressure_system()
    except rospy.ROSInterruptException:
        pass
but error:
[ERROR] [1721475813.497080870]: client wants service /ur_hardware_interface/set_io to have md5sum eb3840e5f3632fc236409b92a9250f5b, but it has f539fc0b1a42859fb186a5aaba22d4b0. Dropping connection.

> I've added another service, I didn't modify the existing one. Therefore you should do
> 
> ```
> self.set_analog_output_client = rospy.ServiceProxy('/ur_hardware_interface/set_analog_output', SetAnalogOutput)
> ```

Hello, I downloaded the driver package from your set_analog branch, but I failed to compile it (I compiled and installed ur_msgs from the set_analog branch and ur_client_library from the analog domain branch at the same time). The error message is as follows:
[ 94%] Building CXX object Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_node.dir/src/ros/hardware_interface.cpp.o
[ 94%] Building CXX object Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_node.dir/src/ros/hardware_interface_node.cpp.o
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp: In lambda function: /home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:394:39: error: ‘ur_msgs::SetPayload::Request’ {aka ‘struct ur_msgs::SetPayloadRequest_<std::allocator<void> >’} has no member named ‘payload’ 394 | << " set_payload(" << req.payload << ", [" << req.center_of_gravity.x << ", " << req.center_of_gravity.y | ^~~~~~~ /home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp: In lambda function: /home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:394:39: error: ‘ur_msgs::SetPayload::Request’ {aka ‘struct ur_msgs::SetPayloadRequest_<std::allocator<void> >’} has no member named ‘payload’ 394 | << " set_payload(" << req.payload << ", [" << req.center_of_gravity.x << ", " << req.center_of_gravity.y | ^~~~~~~ make[2]: *** [Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_node.dir/build.make:76: Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_node.dir/src/ros/hardware_interface.cpp.o] Error 1 make[2]: *** Waiting for unfinished tasks .... make[2]: *** [Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_plugin.dir/build.make:76: Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_plugin.dir/src/ros/hardware_interface.cpp.o] Error 1 make[2]: *** Waiting for unfinished tasks....
[ 95%] Linking CXX executable /home/zp/ur_ws/devel/lib/ur_robot_driver/robot_state_helper
[ 95%] Built target robot_state_helper
make[1]: *** [CMakeFiles/Makefile2:8928: Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_node.dir/all] Error 2
make[1]: *** Waiting for unfinished tasks....
make[1]: *** [CMakeFiles/Makefile2:8284: Universal_Robots_ROS_Driver/ur_robot_driver/CMakeFiles/ur_robot_driver_plugin.dir/all] Error 2
[ 95%] Linking CXX executable /home/zp/ur_ws/devel/lib/ur_robot_driver/dashboard_client [95%] Built target dashboard_client make: *** [Makefile:141:all] Error 2 Invoking "make -j16 -l16" failed

> > I've added another service, I didn't modify the existing one. Therefore you should do
> > ```
> > self.set_analog_output_client = rospy.ServiceProxy('/ur_hardware_interface/set_analog_output', SetAnalogOutput)
> > ```
> 
> Hello, I carefully compared the ur_msgs before and after the change, and found that the definition of SetPayload.srv under the srv folder of the two versions is different.

i solve it,just change the "mass" to "payload", and then i will try again.The compilation passed 100%, although some warnings appeared:
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp: In member function ‘void ur_driver::HardwareInterface::extractToolPose(const ros::Time&)’:
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[3]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  700 |   tcp_transform_.transform.rotation = tf2::toMsg(rotation);
      |                                                          ^
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[2]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[1]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[0]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp: In member function ‘void ur_driver::HardwareInterface::extractToolPose(const ros::Time&)’:
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[3]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  700 |   tcp_transform_.transform.rotation = tf2::toMsg(rotation);
      |                                                          ^
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[2]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[1]’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/ros/hardware_interface.cpp:700:58: warning: ‘rotation.tf2::QuadWord::m_floats[0]’ may be used uninitialized in this function [-Wmaybe-uninitialized]

> I've put together a set of pull requests:
> 
> * [Add a service to set an analog output ros-industrial/ur_msgs#28](https://github.com/ros-industrial/ur_msgs/pull/28)
> * [Analog domain Universal_Robots_Client_Library#211](https://github.com/UniversalRobots/Universal_Robots_Client_Library/pull/211)
> * [Add a set_analog_output service #714](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/714)
> 
> @zp2546265641 If you could try that out and give some feedback, that would be great!
> 
> Please note: When building the client library from source, we suggest removing the installed apt package (`sudo apt remove ros-noetic-ur-client-library`)
Hello, after I compiled successfully, I had a new problem. I started the ur driver file roslaunch ur_robot_driver ur5_bringup.launch robot_ip:=192.168.0.10 kinematics_config:="/home/zp/ur_ws/src/Universal_Robots_ROS_Driver/ur_calibration/etc/ur5_calibration.yaml"


Then the following error appeared, which was not there before.


[ERROR] [1721700688.503446407]: This controller requires a hardware interface of type 'scaled_controllers::ScaledPositionJointInterface'. Make sure this is registered in the hardware_interface::RobotHW class. [ERROR] [1721700688.503503826]: Initializing controller 'scaled_pos_joint_traj_controller' failed [ INFO] [1721700688.505952937]: Robot mode is now RUNNING [ INFO] [1721700688.506612139]: Robot's safety mode is now NORMAL [ERROR] [1721700689.504866]: Failed to load scaled_pos_joint_traj_controller [INFO] [1721700689.506076]: Loading controller: speed_scaling_state_controller [ WARN] [1721700689.513019259]: SEVERE WARNING!!! A namespace collision has occurred with plugin factory for class position_controllers::ScaledJointTrajectoryController. New factory will OVERWRITE existing one. This situation occurs when libraries containing plugins are directly linked against an executable (the one running right now generating this message). Please separate plugins out into their own library or just don't link against the library and use either class_loader::ClassLoader/MultiLibraryClassLoader to open.

--- a/ur_robot_driver/include/ur_robot_driver/hardware_interface.h
+++ b/ur_robot_driver/include/ur_robot_driver/hardware_interface.h
@@ -51,6 +51,7 @@
 #include <ur_client_library/control/trajectory_point_interface.h>
 #include <ur_msgs/IOStates.h>
 #include <ur_msgs/ToolDataMsg.h>
+#include <ur_msgs/SetAnalogOutput.h>
 #include <ur_msgs/SetIO.h>
 #include <ur_msgs/SetSpeedSliderFraction.h>
 #include <ur_msgs/SetPayload.h>
@@ -211,6 +212,7 @@ protected:
 
   bool setSpeedSlider(ur_msgs::SetSpeedSliderFractionRequest& req, ur_msgs::SetSpeedSliderFractionResponse& res);
   bool setIO(ur_msgs::SetIORequest& req, ur_msgs::SetIOResponse& res);
+  bool setAnalogOutput(ur_msgs::SetAnalogOutputRequest& req, ur_msgs::SetAnalogOutputResponse& res);
   bool resendRobotProgram(std_srvs::TriggerRequest& req, std_srvs::TriggerResponse& res);
   bool zeroFTSensor(std_srvs::TriggerRequest& req, std_srvs::TriggerResponse& res);
   void commandCallback(const std_msgs::StringConstPtr& msg);
@@ -319,6 +321,7 @@ protected:
 
   ros::ServiceServer set_speed_slider_srv_;
   ros::ServiceServer set_io_srv_;
+  ros::ServiceServer set_analog_output_srv_;
   ros::ServiceServer resend_robot_program_srv_;
   ros::Subscriber command_sub_;
 

--- a/ur_robot_driver/src/hardware_interface.cpp
+++ b/ur_robot_driver/src/hardware_interface.cpp
@@ -1140,8 +1140,8 @@ bool HardwareInterface::setAnalogOutput(ur_msgs::SetAnalogOutputRequest& req, ur
 {
   if (ur_driver_)
   {
-
-    res.success = ur_driver_->getRTDEWriter().sendStandardAnalogOutput(req.data.pin, req.data.state, static_cast<urcl::AnalogOutputType>(req.data.domain));
+    res.success = ur_driver_->getRTDEWriter().sendStandardAnalogOutput(
+        req.data.pin, req.data.state, static_cast<urcl::AnalogOutputType>(req.data.domain));
   }
   return true;
 }

--- a/ur_robot_driver/src/hardware_interface.cpp
+++ b/ur_robot_driver/src/hardware_interface.cpp
@@ -435,8 +435,9 @@ bool HardwareInterface::init(ros::NodeHandle& root_nh, ros::NodeHandle& robot_hw
   // doing. Using this with other controllers might lead to unexpected behaviors.
   set_speed_slider_srv_ = robot_hw_nh.advertiseService("set_speed_slider", &HardwareInterface::setSpeedSlider, this);
 
-  // Service to set any of the robot's IOs
+  // Services to set any of the robot's IOs
   set_io_srv_ = robot_hw_nh.advertiseService("set_io", &HardwareInterface::setIO, this);
+  set_analog_output_srv_ = robot_hw_nh.advertiseService("set_analog_output", &HardwareInterface::setAnalogOutput, this);
 
   if (headless_mode)
   {
@@ -1132,6 +1133,16 @@ bool HardwareInterface::setIO(ur_msgs::SetIORequest& req, ur_msgs::SetIOResponse
     res.success = false;
   }
 
+  return true;
+}
+
+bool HardwareInterface::setAnalogOutput(ur_msgs::SetAnalogOutputRequest& req, ur_msgs::SetAnalogOutputResponse& res)
+{
+  if (ur_driver_)
+  {
+
+    res.success = ur_driver_->getRTDEWriter().sendStandardAnalogOutput(req.data.pin, req.data.state, static_cast<urcl::AnalogOutputType>(req.data.domain));
+  }
   return true;
 }
 

